### PR TITLE
Enhancement/3288 detect existing tag

### PR DESCRIPTION
**Pull request by eugene-manuilov**
## Summary

Addresses issue #3288

## Checklist

- [ ] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [ ] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

File: assets/js/modules/analytics-4/datastore/tags.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+import Data from 'googlesitekit-data';
+import { createExistingTagStore } from '../../../googlesitekit/data/create-existing-tag-store';
+import { STORE_NAME } from './constants';
+import { tagMatchers } from '../utils/tag-matchers';
+import { isValidWebDataStreamID } from '../utils/validation';
+
+const existingTagStore = createExistingTagStore( {
+	storeName: STORE_NAME,
+	tagMatchers,
+	isValidTag: isValidWebDataStreamID,
```
Comment (felixarntz) [inline assets/js/modules/analytics-4/datastore/tags.js:31]: This shouldn't rely on `isValidWebDataStreamID`, the tag is the "measurement ID" - which is also stored in the web data stream entity, but it's not the web data stream ID.

File: assets/js/modules/analytics-4/utils/validation.js [hunk 2]
```diff
@@ -58,5 +58,5 @@ export function isValidPropertySelection( value ) {
  * @return {boolean} TRUE if the web data stream ID is valid, otherwise FALSE.
  */
 export function isValidWebDataStreamID( webDataStreamID ) {
-	return typeof webDataStreamID === 'string' && /\d+/.test( webDataStreamID );
+	return typeof webDataStreamID === 'string' && /^[a-zA-Z0-9]+$/.test( webDataStreamID );
```
Comment (felixarntz) [inline assets/js/modules/analytics-4/utils/validation.js:61]: See above, the `isValidWebDataStreamID` implementation from before is actually correct. What we need here instead is a new `isValidMeasurementID` function.

File: assets/js/modules/analytics-4/datastore/__factories__/html-with-tag.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+			<head>
+				${ analyticsID && tagHeadHTML( analyticsID ) }
+				<title>Test Title</title>
+			</head>
+			<body>
+				<h1>Test Title</h1>
+			</body>
+		</html>
+	`;
+}
```
Comment (felixarntz) [inline assets/js/modules/analytics-4/datastore/__factories__/html-with-tag.js:52]: Maybe rename the `analyticsID` variables throughout this file to `measurementID` or `analyticsMeasurementID` to clarify what kind of ID is expected.

**Review by felixarntz — CHANGES_REQUESTED** [review]
Mostly lgtm, just a inconsistency that needs to be resolved here which confuses web data stream ID and measurement ID.

File: assets/js/modules/analytics-4/utils/validation.js (hunk 2 above)
Comment (eugene-manuilov) [reply assets/js/modules/analytics-4/utils/validation.js]: Updated.

File: assets/js/modules/analytics-4/datastore/__factories__/html-with-tag.js (hunk 3 above)
Comment (eugene-manuilov) [reply assets/js/modules/analytics-4/datastore/__factories__/html-with-tag.js:52]: Updated.

File: assets/js/modules/analytics-4/utils/tag-matchers.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+export const tagMatchers = [
+	// Detect gtag script calls.
+	/<script [^>]*src=['|"]https:\/\/www.googletagmanager.com\/gtag\/js\?id=G-([a-zA-Z0-9]+)['|"][^>]*><\/script>/i,
```
Comment (tofumatt) [inline assets/js/modules/analytics-4/utils/tag-matchers.js:21]: Technically the unescaped dots here will work, but will arguably match things it shouldn't. Probably won't ever matter, but let's do:

```suggestion
	/<script [^>]*src=['|"]https:\/\/www\.googletagmanager\.com\/gtag\/js\?id=G-([a-zA-Z0-9]+)['|"][^>]*><\/script>/i,
```

**Review by tofumatt — CHANGES_REQUESTED** [review]
Looks good, but let's fix up the Tag Manager regex.

File: assets/js/modules/analytics-4/utils/tag-matchers.js (hunk 4 above)
Comment (eugene-manuilov) [reply assets/js/modules/analytics-4/utils/tag-matchers.js:21]: Yep, agree.

**Review by tofumatt — APPROVED** [review]
Looks great. Not sure why there's an extra merge commit in here from today, but it doesn't seem to be an issue. Feel free to clarify in the PR if you get the chance, @danielgent 🙂

**Pull request merged**